plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -119,31 +120,126 @@
-                refreshInterval
-                        .map(Duration::toMillis)
-                        .map(OptionalLong::of)
-                        .orElseGet(OptionalLong::empty),
-                Optional.of(executor),
-                maximumSize,
-                StatsRecording.ENABLED,
-                partitionCacheEnabled);
+        return new CachingHiveMetastoreBuilder();
     }
 
     public static CachingHiveMetastore memoizeMetastore(HiveMetastore delegate, long maximumSize)
     {
-        return new CachingHiveMetastore(
-                delegate,
-                OptionalLong.empty(),
-                OptionalLong.empty(),
-                Optional.empty(),
-                maximumSize,
-                StatsRecording.DISABLED,
-                true);
+        return builder()
+                .delegate(delegate)
+                .maximumSize(maximumSize)
+                .statsRecording(StatsRecording.DISABLED)
+                .build();
+    }
+
+    @Immutable
+    public static class CachingHiveMetastoreBuilder
+    {
+        private HiveMetastore delegate;
+        private Optional<Executor> executor = Optional.empty();
+        private OptionalLong expiresAfterWriteMillis = OptionalLong.empty();
+        private OptionalLong refreshMills = OptionalLong.empty();
+        private Long maximumSize;
+        private StatsRecording statsRecording = StatsRecording.ENABLED;
+        private boolean partitionCacheEnabled = true;
+
+        public CachingHiveMetastoreBuilder() {}
+
+        private CachingHiveMetastoreBuilder(
+                HiveMetastore delegate,
+                Optional<Executor> executor,
+                OptionalLong expiresAfterWriteMillis,
+                OptionalLong refreshMills,
+                Long maximumSize,
+                StatsRecording statsRecording,
+                boolean partitionCacheEnabled)
+        {
+            this.delegate = delegate;
+            this.executor = executor;
+            this.expiresAfterWriteMillis = expiresAfterWriteMillis;
+            this.refreshMills = refreshMills;
+            this.maximumSize = maximumSize;
+            this.statsRecording = statsRecording;
+            this.partitionCacheEnabled = partitionCacheEnabled;
+        }
+
+        public CachingHiveMetastoreBuilder delegate(HiveMetastore delegate)
+        {
+            CachingHiveMetastoreBuilder builder = duplicate();
+            builder.delegate = requireNonNull(delegate, "delegate is null");
+            return builder;
+        }
+
+        public CachingHiveMetastoreBuilder executor(Executor executor)
+        {
+            CachingHiveMetastoreBuilder builder = duplicate();
+            builder.executor = Optional.of(requireNonNull(executor, "executor is null"));
+            return builder;
+        }
+
+        public CachingHiveMetastoreBuilder cacheTtl(Duration cacheTtl)
+        {
+            CachingHiveMetastoreBuilder builder = duplicate();
```

Comment:
i like the immutability of this builder, but I am concerned this is now how most of the builders are implemented, and may be confusing to users.
i.e. if i write code that just calls `builder.cacheTtl(..);`, i doubt a reviewer would catch the mistake.

thoughts?

we can at least help a bit annotating the builder methods with `@CheckReturnValue`. A reviewer may not see that, but tooling may

Reply:
The other option I was thinking was adding `CachingHiveMetastoreBuilder#build(HiveMetastore delegate, Executor executor)` method (these fields changed in `io.trino.plugin.hive.metastore.cache.SharedHiveMetastoreCache#createCachingHiveMetastoreFactory`)?

wdyt?

Reply:
@losipiuk thoughts?

Reply:
addressed with `CachingHiveMetastoreBuilder builder(CachingHiveMetastoreBuilder other)` method